[Asset Inventory] Fix URL query handling for asset inventory Flyout

opauloh opened the pull request:
## Summary


It fixes https://github.com/elastic/security-team/issues/12544

This PR updates the URL query handling logic to allow Rison encoding and decoding for multiple keys. This fixes an issue where reloading the page or filtering/searching were making the flyout disappear since it was missing it's reference on the URL params.


* **Added support for flyout and multiple Rison parameters**: New functions (`encodeFlyout`, `decodeFlyout`, `encodeRisonParam`, `decodeRisonParam`, `encodeMultipleRisonParams`, `decodeMultipleRisonParams`) were introduced to handle encoding and decoding of flyout parameters and multiple Rison parameters. 

### Refactoring of `useUrlQuery` hook:

* **Support for multiple URL parameters**: The `useUrlQuery` hook was refactored to decode and encode multiple Rison parameters, including flyout-specific parameters, while maintaining separation between query and flyout namespaces. Helper functions (`buildFlyoutSearchString`, `combineSearchParts`) were added to streamline the construction of search strings.

### Recording


https://github.com/user-attachments/assets/14205787-48dd-4306-bc0f-ac25958b694a



<!--ONMERGE {"backportTargets":["9.1"]} ONMERGE-->

opauloh committed:
Refactors URL query handling for asset inventory

Updates the URL query handling logic to leverage rison encoding for improved query parameter management.

This change introduces a new hook, `useUrlQuery`, to handle URL parameters for the asset inventory. It uses `encodeQuery` for main query parameters and `encodeRisonParam` for flyout parameters, ensuring proper encoding.

The hook provides functions to get and set URL query parameters, allowing for seamless integration with the asset inventory components.

Adds a test suite for the `useUrlQuery` hook.

opauloh committed:
Adds Rison encoding/decoding utilities

Introduces utilities for encoding and decoding data using Rison format for URL parameters.

This provides functions to:
- Encode and decode single and multiple Rison parameters.
- Manage flyout parameters within URLs.

These utilities enhance the ability to serialize and deserialize complex data structures within URLs, improving state management and navigation within the application.

opauloh committed:
Adds utility functions for query parameter handling

Introduces a new module with utility functions for encoding and decoding query parameters using Rison.

This provides a consistent and reusable way to manage URL parameters for the cloud security posture feature, including functions for single and multiple parameters as well as generic encoding/decoding.

opauloh committed:
Refactors URL query parameter handling

Updates URL query parameter handling to use constants for query parameter keys, improving maintainability and readability.

Consolidates flyout parameter extraction and encoding into helper functions for clarity.

opauloh committed:
Merge branch 'main' into asset-inventory/fix-url-query-issue

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/312449)
* Commit: 48178994756db373f726194c04ee98f7af9b7875

### Failed CI Steps
* [Linting](https://buildkite.com/elastic/kibana-pull-request/builds/312449#0197a578-2bd5-4a57-b637-9b129f93dbda)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/312449","number":312449,"commit":"48178994756db373f726194c04ee98f7af9b7875"}],"number":312449}
buildkite-pr-comment-->

opauloh committed:
Fixes return type for query encoding

Updates the return type of the query encoding function to align with the expected type definition.
This ensures that the function returns a value compatible with the location descriptor object's search property, resolving a type mismatch.

opauloh committed:
Merge branch 'main' into asset-inventory/fix-url-query-issue

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/313317)
* Commit: 84a875496e4e3b3700c6c29bba4d8040d50811b2
* [Cloud Deployment](https://kibana-pr-225199.kb.us-west2.gcp.elastic-cloud.com)

### History
* :broken_heart: [Build #312449](https://buildkite.com/elastic/kibana-pull-request/builds/312449) failed 48178994756db373f726194c04ee98f7af9b7875

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/313317","number":313317,"commit":"84a875496e4e3b3700c6c29bba4d8040d50811b2"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/312449","number":312449,"commit":"48178994756db373f726194c04ee98f7af9b7875"}],"number":313317}
buildkite-pr-comment-->

opauloh committed:
Updates mocked query string

Updates the mocked query string in a test file to match the expected `cspq` parameter name.

This ensures the test accurately simulates the expected URL structure and avoids potential mismatches during query string encoding and decoding.

opauloh committed:
Merge remote-tracking branch 'origin/asset-inventory/fix-url-query-issue' into asset-inventory/fix-url-query-issue

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/313322)
* Commit: 1f76873f0fa97d28fc06703cfd2e0b0f933de608
* [Cloud Deployment](https://kibana-pr-225199.kb.us-west2.gcp.elastic-cloud.com)

### History
* :broken_heart: [Build #313317](https://buildkite.com/elastic/kibana-pull-request/builds/313317) failed 84a875496e4e3b3700c6c29bba4d8040d50811b2
* :broken_heart: [Build #312449](https://buildkite.com/elastic/kibana-pull-request/builds/312449) failed 48178994756db373f726194c04ee98f7af9b7875

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/313322","number":313322,"commit":"1f76873f0fa97d28fc06703cfd2e0b0f933de608"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/313317","number":313317,"commit":"84a875496e4e3b3700c6c29bba4d8040d50811b2"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/312449","number":312449,"commit":"48178994756db373f726194c04ee98f7af9b7875"}],"number":313322}
buildkite-pr-comment-->

animehart reviewed (approved):
lgtm !

opauloh committed:
Merge branch 'main' into asset-inventory/fix-url-query-issue

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/313352)
* Commit: 88d2164e2fd1a84b885fa704e7179d58c2794809

### Failed CI Steps
* [Build and Deploy to Cloud](https://buildkite.com/elastic/kibana-pull-request/builds/313352#0197b04f-2c93-4a90-b385-9891b4402d51)

### History
* :green_heart: [Build #313322](https://buildkite.com/elastic/kibana-pull-request/builds/313322) succeeded 1f76873f0fa97d28fc06703cfd2e0b0f933de608
* :broken_heart: [Build #313317](https://buildkite.com/elastic/kibana-pull-request/builds/313317) failed 84a875496e4e3b3700c6c29bba4d8040d50811b2
* :broken_heart: [Build #312449](https://buildkite.com/elastic/kibana-pull-request/builds/312449) failed 48178994756db373f726194c04ee98f7af9b7875

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/313352","number":313352,"commit":"88d2164e2fd1a84b885fa704e7179d58c2794809"},{"buildStatus":{"state":"running","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/313322","number":313322,"commit":"1f76873f0fa97d28fc06703cfd2e0b0f933de608"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/313317","number":313317,"commit":"84a875496e4e3b3700c6c29bba4d8040d50811b2"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/312449","number":312449,"commit":"48178994756db373f726194c04ee98f7af9b7875"}],"number":313352}
buildkite-pr-comment-->

kibanamachine commented:
Starting backport for target branches: 9.1

https://github.com/elastic/kibana/actions/runs/15931922721

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|9.1|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/225686">](https://github.com/elastic/kibana/pull/225686)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->
